### PR TITLE
[3Box Integration] Add 'Upload to 3Box' button to Avatar Builder

### PR DESCRIPTION
**Pull request by think-in-universe**
##### Description

Based on the feature requested for https://github.com/3box/3box/issues/839, implement the GitCoin-3Box integration feature for Avatar Builder per the request of Kevin (@owocki). 

We will take advantage of what we have achieved in GitCoin and 3Box integration in https://github.com/gitcoinco/web/pull/5640, and extend the backup functionalities to Avatar Builder.

The design doc for this feature can be found here: https://hackmd.io/@L9ZnGph8Tbaio2GVX_z-EA/SylymWCAr

##### Refers/Fixes

https://github.com/gitcoinco/web/issues/5660

##### Testing

Preview for this feature can be found below:

1. "upload to 3box" button

image

image

2. Verify the uploaded custom avatar data in 3Box Hub

image

(looks a bit messy due to the 3Box Hub UI)

image

### CODE REVIEW

**Comment by Sirlupinwatson1** [comment]
Nice Project @think-in-universe  i will give a new face to @owocki  haha

**Comment by think-in-universe** [comment]
@Sabihashaik Haha. Awesome!!

I believe avatars are easier to be shared and collaboratively edited if it can be made open with data infrastructure like 3Box.

**Comment by owocki** [comment]
@think-in-universe awesome, im pumped for this!

kicking it over to @gitcoinco/engineers for review..  i believe that john  from ethdenver wants this live before ETHDenver on 2/13

File: app/assets/v2/js/3box-sync.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    });
+  }
+
+  async function saveDataToSpace(space, model) {
+    const res = await fetchProfieData(model);
+
+    if (res.data) {
+      let data = res.data;
+
+      console.log('data', data);
```
Comment (thelostone-mc) [inline app/assets/v2/js/3box-sync.js]: can we remove the log?

File: app/assets/v2/js/3box-sync.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    const unused_public_keys = Object.keys(public_data).filter(x => !public_keys.includes(x));
+    const unused_private_keys = Object.keys(private_data).filter(x => !private_keys.includes(x));
+
+    await removeFields(space.public, unused_public_keys);
+    await removeFields(space.private, unused_private_keys);
+
+    const count = unused_public_keys.length + unused_private_keys.length;
+
+    console.log(`remove ${count} outdated fields from space`, unused_public_keys, unused_private_keys);
```
Comment (thelostone-mc) [inline app/assets/v2/js/3box-sync.js]: Do we need L113 , L115

Comment (think-in-universe) [reply app/assets/v2/js/3box-sync.js]: not necessary. let me remove the logs

File: app/assets/v2/js/3box-sync.js (hunk 1 above)
Comment (think-in-universe) [reply app/assets/v2/js/3box-sync.js]: removed

**Comment by think-in-universe** [comment]
@thelostone-mc thanks. the logs are removed.

**Comment by octavioamu** [comment]
we are merging #5947 that have a lot of duplicate code.
So this need a rebase/merge and will need also the migrations to be recreated.
Can you take a look @think-in-universe ?

**Comment by think-in-universe** [comment]
@octavioamu sure. I could help with merge and update migrations after we fix the issue with https://github.com/gitcoinco/web/pull/5955 . 

I understand we'd like to have the feature ready before Feb 13th, so I'll help with the code clean up and more testing once https://github.com/gitcoinco/web/pull/5955 is resolved.

**Comment by think-in-universe** [comment]
@octavioamu merged with master branch, and tested for saving profile and avatars to 3box locally, both works well.

**Comment by think-in-universe** [comment]
please note that `3box.min.js` is now included via ```<script src="https://unpkg.com/3box/dist/3box.min.js"></script>``` in `onboard.html` and `profile.html`

please let me know if we prefer add `3box.min.js` it into `app/assets/v2/js/lib/`

**Comment by think-in-universe** [comment]
resolved the merge conflict

**Comment by think-in-universe** [comment]
could anyone take a look at the test failure of `DashboardUtilsTest.test_get_bounty` in travis? looks like a network error?

**Comment by Sirlupinwatson1** [comment]
> could anyone take a look at the test failure of `DashboardUtilsTest.test_get_bounty` in travis? looks like a network error?

requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://rinkeby.infura.io/

Also incompatible version error.

**Comment by think-in-universe** [comment]
@Sirlupinwatson1 thanks~ is that an environment issue or related to my code change?

**Comment by Sirlupinwatson1** [comment]
> @Sirlupinwatson1 thanks~ is that an environment issue or related to my code change?

This is what i can read : 

       1 failed
         - app/dashboard/tests/test_dashboard_utils.py:62 DashboardUtilsTest.test_get_bounty
The command "pytest -p no:ethereum" exited with 1.

So this is a environment issue to me ? I can't guess on a answer. @danlipert did advertise me to add concrete commente so i don't want to confuse you on your work. I just wanted to help.

**Comment by think-in-universe** [comment]
@Sirlupinwatson1 thanks. 😄 

I rerun `make pytest` locally and all tests (including `DashboardUtilsTest.test_get_bounty`) passed. I think this might be a random failure. 

And anyway the python test should not related to the change in the commit https://github.com/gitcoinco/web/pull/5667/commits/bbdaa14fbbba6510da7c11828b6e5f50086e10fc, which only made one-line change in html for resolving merge conflict.

**Comment by Sirlupinwatson1** [comment]
> @Sirlupinwatson1 thanks. 😄
> 
> I rerun `make pytest` locally and all tests (including `DashboardUtilsTest.test_get_bounty`) passed. I think this might be a random failure.
> 
> And anyway the python test should not related to the change in the commit [bbdaa14](https://github.com/gitcoinco/web/commit/bbdaa14fbbba6510da7c11828b6e5f50086e10fc), which only made one-line change in html for resolving merge conflict.

I have this question in my mind, not sure if i can ask it here. But in the log files, there is a message saying about Interference for the TimeZone -8 ,  these warnings should be ignored ? I know also this is just a integration test but to be a perfect code we should not ignore those warnings ? 

I would be happy if someone could answer to this question about the timezone log error, but also to give the right answer about the purpose of the main error it was randomly giving like @think-in-universe is saying.

Have a good day @think-in-universe !

**Comment by think-in-universe** [comment]
@Sirlupinwatson1 thanks~

@owocki hey Kevin, if you want to get ready for ETHDenver is to backup the avatar builder data first, I think we could merge this PR first which will only save the `custom_avatar` fields into 3box in the avatar builder page, rather than save much more fields in the profile backup feature in the profile page.

**Comment by owocki** [comment]
thanks.. i dont have the cycles to dive into the weeds on this tho :-/

**Comment by octavioamu** [comment]
@owocki @think-in-universe I will test it locally

**Comment by octavioamu** [comment]
image
I don't see it on my 3box account but I believe is because the image is on localhost that is right?

**Comment by octavioamu** [comment]
Never mind I found it. Working perfectly  
image

**Comment by octavioamu** [comment]
> > @Sirlupinwatson1 thanks. 😄
> > I rerun `make pytest` locally and all tests (including `DashboardUtilsTest.test_get_bounty`) passed. I think this might be a random failure.
> > And anyway the python test should not related to the change in the commit [bbdaa14](https://github.com/gitcoinco/web/commit/bbdaa14fbbba6510da7c11828b6e5f50086e10fc), which only made one-line change in html for resolving merge conflict.
> 
> I have this question in my mind, not sure if i can ask it here. But in the log files, there is a message saying about Interference for the TimeZone -8 , these warnings should be ignored ? I know also this is just a integration test but to be a perfect code we should not ignore those warnings ?
> 
> I would be happy if someone could answer to this question about the timezone log error, but also to give the right answer about the purpose of the main error it was randomly giving like @think-in-universe is saying.
> 
> Have a good day @think-in-universe !

that error is not from this pr actually @Sirlupinwatson1 is failing on travis in stable and master so we should check that separately from this review.

**Comment by think-in-universe** [comment]
@octavioamu cool. 

is it necessary to merge this PR first when we're fixing #5955 with #5981?

**Comment by octavioamu** [comment]
> @octavioamu cool.
> 
> is it necessary to merge this PR first when we're fixing #5955 with #5981?

not sure you tell me!! But the idea is having the 2 uploads for ethdenver

**Comment by Sirlupinwatson1** [comment]
> > > @Sirlupinwatson1 thanks. 😄
> > > I rerun `make pytest` locally and all tests (including `DashboardUtilsTest.test_get_bounty`) passed. I think this might be a random failure.
> > > And anyway the python test should not related to the change in the commit [bbdaa14](https://github.com/gitcoinco/web/commit/bbdaa14fbbba6510da7c11828b6e5f50086e10fc), which only made one-line change in html for resolving merge conflict.
> > 
> > 
> > I have this question in my mind, not sure if i can ask it here. But in the log files, there is a message saying about Interference for the TimeZone -8 , these warnings should be ignored ? I know also this is just a integration test but to be a perfect code we should not ignore those warnings ?
> > I would be happy if someone could answer to this question about the timezone log error, but also to give the right answer about the purpose of the main error it was randomly giving like @think-in-universe is saying.
> > Have a good day @think-in-universe !
> 
> that error is not from this pr actually @Sirlupinwatson1 is failing on travis in stable and master so we should check that separately from this review.

Thanks for the answer @octavioamu , much appreciate. 

From what i understand we need to update out things up and open-up another issue, but since you are all working on this issue and there is a rush , you will not open up a issue and just do it ?

I know this is running out of what is relevant here but i really do need to learn this and know why since i am learning to code. Thanks again.

**Comment by think-in-universe** [comment]
@octavioamu sure. I think we're mainly blocked for testing with the profile data in production. 

as I mentioned in #5981 , it would be nice if we could run more testing in alpha, in order to collect the call stack (if met 500 error when export data), and query timeframe for large size profile.

**Comment by think-in-universe** [comment]
@octavioamu should we also merge this feature after #5981 is well tested?

**Pull request merged**